### PR TITLE
Support custom LSA auth provider

### PR DESCRIPTION
**Pull request by bagajjal**
1) Support custom LSA auth provider (username;password;domain format).
2) Fix the _dupenv_s issue (length variable is passed to _dupenv_s).

### CODE REVIEW

File: auth-passwd.c [hunk 1]
```diff
@@ -256,18 +304,27 @@ int sys_auth_passwd(Authctxt *authctxt, const char *password)
...
-		else
-			debug("failed to logon user: %ls domain: %ls error:%d", user_utf16, udom_utf16, GetLastError());
+		else {
+			debug("Windows authentication failed for user: %ls domain: %ls error:%d", user_utf16, udom_utf16, GetLastError());
+
+			/* If LSA authentication package is configured then it will return the auth_token */
+			sys_auth_passwd_lsa(authctxt, password);
+		}
+			
 		goto done;
```
Comment (bagajjal) [inline auth-passwd.c]: Manoj - reorder the code to remove the goto done as it is not so readable.

File: auth-passwd.c [hunk 2]
```diff
@@ -59,6 +56,11 @@
 #include "auth-options.h"
 #include "authfd.h"
 
+#ifdef WINDOWS
+#include <logonuser.h>
```
Comment (bingbing8) [inline auth-passwd.c]: should be "logonuser.h" instead of <logonuser.h>?
same as pwd.h?

File: auth-passwd.c [hunk 3]
```diff
@@ -228,10 +230,56 @@ sys_auth_passwd(Authctxt *authctxt, const char *password)
...
+				domain = (char*) malloc(domain_len * sizeof(char));
+				memcpy(domain, tmp, domain_len);
+				domain[domain_len] = '\0';
+			}
+			debug("Authenticating using LSA Auth Package:%ls", lsa_auth_pkg_w);
+			authctxt->auth_token = mm_auth_custom_lsa(authctxt->pw->pw_name, password, domain, lsa_auth_pkg);
+		}
+	}
+
+done:
```
Comment (bingbing8) [inline auth-passwd.c:265]: Need to call RegCloseKey to close the handle of reg_key

File: contrib/win32/win32compat/win32_monitor_wrap.c [hunk 4]
```diff
@@ -211,4 +211,44 @@ int mm_load_profile(const char* user_name, u_int token)
...
+		if (!msg)
+			fatal("%s: out of memory", __func__);
+
+		if (sshbuf_put_u8(msg, SSH_PRIV_AGENT_MSG_ID) != 0 ||
+			sshbuf_put_cstring(msg, CUSTOM_LSA_AUTH_REQUEST) != 0 ||
+			sshbuf_put_cstring(msg, user) != 0 ||
+			sshbuf_put_cstring(msg, password) != 0 ||
+			sshbuf_put_cstring(msg, domain) != 0 ||
+			sshbuf_put_cstring(msg, lsa_auth_pkg) != 0 ||
+			ssh_request_reply(agent_fd, msg, msg) != 0) {
```
Comment (bingbing8) [inline contrib/win32/win32compat/win32_monitor_wrap.c:243]: for troubleshooting purpose, can we separate the sh_request_reply from the sending and log debug info?

File: contrib/win32/win32compat/priv-agent.h [hunk 5]
```diff
@@ -6,3 +6,4 @@
 #define SSH_PRIV_AGENT_MSG_ID			0
 #define PUBKEY_AUTH_REQUEST			"pubkey"
 #define LOAD_USER_PROFILE_REQUEST		"loadprofile"
+#define CUSTOM_LSA_AUTH_REQUEST			"lsaauth"
```
Comment (bingbing8) [inline contrib/win32/win32compat/priv-agent.h]: fix the white spaces to align with above statements

File: contrib/win32/win32compat/ssh-agent/authagent-request.c [hunk 6]
```diff
@@ -250,6 +252,128 @@ duplicate_token_for_client(struct agent_connection* con, HANDLE t) {
...
+		&subStatus)) != STATUS_SUCCESS) {		
+		if(ret == STATUS_ACCOUNT_RESTRICTION)
+			error("LsaLogonUser failed, error:%ld subStatus:%ld", LsaNtStatusToWinError(ret), subStatus);
+		else
+			error("LsaLogonUser failed error:%ld", LsaNtStatusToWinError(ret));
+
+		goto done;
+	}
+
+	if ((dup_token = duplicate_token_for_client(con, token)) == NULL)
```
Comment (bingbing8) [inline contrib/win32/win32compat/ssh-agent/authagent-request.c:348]: if we are sure finishing useing dup_token, better to close it here

**Review by bingbing8 — CHANGES_REQUESTED** [review]

File: contrib/win32/win32compat/ssh-agent/authagent-request.c (hunk 6 above)
Comment (bagajjal) [reply contrib/win32/win32compat/ssh-agent/authagent-request.c:348]: duplicate token is sent to the sshd.. 
closing the token here doesn't have any impact as the duplicate token is registered under client/sshd process..

File: auth-passwd.c (hunk 3 above)
Comment (bagajjal) [reply auth-passwd.c:265]: fixed.

File: auth-passwd.c (hunk 2 above)
Comment (bagajjal) [reply auth-passwd.c]: fixed "logonuser.h"..
pwd.h is a common header file so not fixing it.

File: contrib/win32/win32compat/win32_monitor_wrap.c (hunk 4 above)
Comment (bagajjal) [reply contrib/win32/win32compat/win32_monitor_wrap.c:243]: fixed

Comment (bingbing8) [reply contrib/win32/win32compat/win32_monitor_wrap.c]: ssh_request_reply is still in the same if condition

File: contrib/win32/win32compat/win32_monitor_wrap.c [hunk 7]
```diff
@@ -211,4 +216,44 @@ int mm_load_profile(const char* user_name, u_int token)
...
+			debug("unable to send LSA Auth request");
+			break;
+		}
+
+		if (sshbuf_get_u32(msg, (u_int32_t *) &token) != 0)
```
Comment (bingbing8) [inline contrib/win32/win32compat/win32_monitor_wrap.c]: can you add a debug trace here?

**Review by bingbing8 — APPROVED** [review]

**Review by manojampalam — APPROVED** [review]
Reviewed offline

**Pull request merged**